Add CacheStore connection pool docs to guide (#37393)
rails/rails: guides/source/caching_with_rails.md

* `:race_condition_ttl` - This option is used in conjunction with the `:expires_in` option. It will prevent race conditions when cache entries expire by preventing multiple processes from simultaneously regenerating the same entry (also known as the dog pile effect). This option sets the number of seconds that an expired entry can be reused while a new value is being regenerated. It's a good practice to set this value if you use the `:expires_in` option.

> When using Puma or another threaded server it is recommended to use a connection pool to prevent thread contention. Connection pooling is only supported by the `MemCacheStore` and `RedisCacheStore`. To enable the connection pool first add the `connection_pool` gem to your Gemfile.

Reviewer: > it is recommended to use a connection pool to prevent thread contention

This won't mean anything to the reader unless they already understand what a connection pool does. Googling "thread contention" is likely to scare off anyone without a computer science background.

I don't think we need to go quite this far, but the Configuring Rails Applications guide has a more detailed explanation of why Active Record's connection pool exists and how it works:

https://github.com/rails/rails/blob/8eef8791fab0985060e7a17687698383b3d47c2c/guides/source/configuring.md#database-pooling

> When using Puma or another threaded server it is recommended to use a connection pool to prevent thread contention. Connection pooling is only supported by the `MemCacheStore` and `RedisCacheStore`. To enable the connection pool first add the `connection_pool` gem to your Gemfile.

Author: I've moved the `connection pool options` to a separate subsection, made the suggested changes, and changed the `thread contention` recommendation to a suggestion:
```
By default the `MemCacheStore` and `RedisCacheStore` use a single connection
per process. This means that if you're using Puma, or another threaded server,
you can have multiple threads waiting for the connection to become available.
To increase the amount of available connections you can enable connection
pooling.
```

```ruby
gem 'connection_pool'
```

> Next set either the `:pool_size` and/or `:pool_timeout` options.

Reviewer: ```suggestion
Next, set the `:pool_size` and/or `:pool_timeout` options:
```

I think this could use a code block example:

```ruby
config.cache_store = :mem_cache_store, "cache.example.com", pool_size: 5, pool_timeout: 5
```

Author: I've made the suggested changes and added the code example.

> * `:pool_size` - This options sets the number of connections per process. It is recommended to set it to the number of threads per process of the app server.

Reviewer: > It is recommended

By whom? 😄 depending on your configuration, this could flip the failure mode from "requests get slower when cache operations are slow" to the potentially disastrous "many more cache connections are opened when cache operations are slow".

I don't have a concrete suggestion here, but notice how the Configuring Rails Applications guide avoids prescribing a particular configuration while still giving some guidance:

https://github.com/rails/rails/blob/8eef8791fab0985060e7a17687698383b3d47c2c/guides/source/configuring.md#database-pooling

> If you get the above error, you might want to increase the size of the connection pool by incrementing the `pool` option in `database.yml`

> * `:pool_timeout` - This option sets the amount of time in seconds after which a connection times out. If no connection is available within timeout seconds it will raise a Timeout::Error.

Reviewer: The first sentence currently sounds like it describes an idle timeout, rather than a checkout timeout.

```suggestion
* `:pool_timeout` - This option sets the amount of time in seconds to wait for a connection. If no connection is available within the timeout, a `Timeout::Error` will be raised.
```

Author: Good point! 🙂 I Actually thought that was how it works.
I've made the suggested changes (and added the default values as well).

* `:race_condition_ttl` - This option is used in conjunction with the `:expires_in` option. It will prevent race conditions when cache entries expire by preventing multiple processes from simultaneously regenerating the same entry (also known as the dog pile effect). This option sets the number of seconds that an expired entry can be reused while a new value is being regenerated. It's a good practice to set this value if you use the `:expires_in` option.

#### Connection Pool Options

By default the `MemCacheStore` and `RedisCacheStore` use a single connection
per process. This means that if you're using Puma, or another threaded server,
you can have multiple threads waiting for the connection to become available.

> To increase the amount of available connections you can enable connection

Reviewer: ```suggestion
To increase the number of available connections you can enable connection
```

Since connections are countable, we should use "number of" here.

https://english.stackexchange.com/a/9442

pooling. Connection pooling is only supported by the `MemCacheStore` and

> `RedisCacheStore`.

Reviewer: > Connection pooling is only supported by the `MemCacheStore` and `RedisCacheStore`.

I think this is implied when they're listed in the first sentence in the paragraph, no need to repeat it here.

Author: I've applied the suggested changes.

> To enable the connection pool first add the `connection_pool` gem to your

Reviewer: The preceding paragraph now mentions that we're enabling connection pooling, so I think we can shorten this:

```suggestion
First, add the `connection_pool` gem to your Gemfile:
```

Author: I've applied the suggested changes.

Gemfile:

```ruby
gem 'connection_pool'
```

> Next, set the `:pool_size` and/or `:pool_timeout` options:

Reviewer: ```suggestion
Next, pass the `:pool_size` and/or `:pool_timeout` options when configuring the cache store:
```

Author: I've applied the suggested changes.

```ruby
config.cache_store = :mem_cache_store, "cache.example.com", pool_size: 5, pool_timeout: 5
```

* `:pool_size` - This option sets the number of connections per process (defaults to 5).

> * `:pool_timeout` - This option sets the amount of time in seconds to wait for a connection (defaults to 5). If no connection is available within the timeout, a `Timeout::Error` will be raised .

Reviewer: ```suggestion
* `:pool_timeout` - This option sets the number of seconds to wait for a connection (defaults to 5). If no connection is available within the timeout, a `Timeout::Error` will be raised.
```

"Amount of time in seconds" is grammatically correct but a bit clunky; we already use "number of seconds" elsewhere in this guide:

https://github.com/rails/rails/blob/8ae8e19a33f4749e529b1649f4f0ace1dbb37285/guides/source/caching_with_rails.md#L371

There's also an extra space before the final full stop here ✂️

Author: I've applied the suggested changes.
